alns/select/MABSelector.py

```diff
@@ -0,0 +1,201 @@
+    num_destroy
+        Number of destroy operators.
+    num_repair
+        Number of repair operators.
+    op_coupling
+        Optional boolean matrix that indicates coupling between destroy and
+        repair operators. Entry (i, j) is True if destroy operator i can be
+        used together with repair operator j, and False otherwise.
+    context_extractor
+        Optional function that takes a ALNS State object and returns a context
+        vector for that state that can be passed to a contextual MABWiser
+        bandit. If the MAB algorithm supports it, this context will be used to
+        help predict the next (destroy, repair) combination.
+
+    References
+    ----------
+    .. [1] Emily Strong, Bernard Kleynhans, & Serdar Kadioglu (2021).
+           MABWiser: Parallelizable Contextual Multi-armed Bandits.
+           Int. J. Artif. Intell. Tools, 30(4), 2150021:1–2150021:19.
+    """
+
+    def __init__(
+        self,
+        scores: List[float],
+        num_destroy: int,
+        num_repair: int,
+        learning_policy: "LearningPolicy",
+        neighborhood_policy: Optional["NeighborhoodPolicy"] = None,
+        seed: Optional[int] = None,
+        op_coupling: Optional[np.ndarray] = None,
+        **kwargs,
+    ):
+        if not MABWISER_AVAILABLE:
+            raise ImportError("MABSelector requires the MABWiser library. ")
+
+        super().__init__(num_destroy, num_repair, op_coupling)
+
+        if any(score < 0 for score in scores):
+            raise ValueError("Negative scores are not understood.")
+
+        if len(scores) < 4:
+            # More than four is OK because we only use the first four.
+            raise ValueError(f"Expected four scores, found {len(scores)}")
+
+        # forward the seed argument if not null
+        if seed is not None:
+            kwargs["seed"] = seed
+
+        # the set of valid operator pairs (arms) is equal to the cartesian
+        # product of destroy and repair operators, except we leave out any
+        # pairs disallowed by op_coupling
+        arms = [
+            f"{d_idx}_{r_idx}"
+            for d_idx, r_idx in itertools.product(
+                range(num_destroy), range(num_repair)
+            )
+            if op_coupling is None or op_coupling[d_idx, r_idx]
+        ]
+        self._mab = MAB(
+            arms,
+            learning_policy,
+            neighborhood_policy,
+            **kwargs,
+        )
+        self._scores = scores
+
+        def extract_context(state: ContextualState):
+            if self._mab.is_contextual is False:
+                return None
+            else:
+                try:
+                    return np.atleast_2d(state.get_context())
+                except AttributeError:
+                    raise ValueError(
+                        "Contextual MAB requires a `get_context` function on"
+                        " the state object"
+                    )
```

Comment:
```suggestion
                return np.atleast_2d(state.get_context())
```
The types already suggest the method should exist. There's no need to check that again - if it fails, the AttributeError is already appropriate.